src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.cs

```diff
@@ -0,0 +1,267 @@
+    ///
+    /// Several constraints are applied before we suggest modifying a method signature (either one of its parameters or its return type):
+    ///
+    ///   * The method cannot be implementing any interface.
+    ///
+    ///   * The method cannot be virtual, abstract, or be an override.
+    ///
+    ///   * The method must be private.
+    ///
+    ///   * The method must not have been assigned to a delegate.
+    /// </remarks>
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed partial class UseConcreteTypeAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1859";
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForMethodReturn = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForMethodReturnMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForParameter = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForParameterMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForLocal = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForLocalMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        internal static readonly DiagnosticDescriptor UseConcreteTypeForField = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeTitle)),
+            CreateLocalizableResourceString(nameof(UseConcreteTypeForFieldMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(UseConcreteTypeDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(UseConcreteTypeForField);
```

Comment:
We normally pass in all descriptors here, even if they share the same rule ID.

Reply:
Note there are meta-analyzers that will report warnings if code with an analyzer uses a descriptor to report a diagnostic but that descriptor is not part of the array returned in SupportedDiagnostics.